Выполнить ТЗ №12

SergeyLeshchev opened the pull request:
1. Поменять реализацию на работу с БД
2. Добавить мапперы
3. Изменить модели согласно новым требованиям
4. Добавить дополнительные тесты
5. Добавить генерацию БД и таблиц
6. Заменить bddiagram

SergeyLeshchev committed:
Выполнить ТЗ №12

1. Поменять реализацию на работу с БД
2. Добавить мапперы
3. Изменить модели согласно новым требованиям
4. Добавить дополнительные тесты
5. Добавить генерацию БД и таблиц
6. Заменить bddiagram

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/dal/mappers/LikesRowMapper.java:
```diff
@@ -0,0 +1,17 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+@Component("likesRowMapper")
```
А зачем названия для компонентов? Мапперы вообще можно не делать компонентами, а просто реализовать эти мап-функции в самих стораджах

SergeyLeshchev replied:
У меня Spring не мог правильно найти и внедрить бины, поэтому я помечал чуть ли не всё, чтобы работало. Попробую убрать лишние и проверять, чтобы не сломалось)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/dal/storage/BaseStorage.java:
```diff
@@ -0,0 +1,55 @@
+    protected final RowMapper<T> mapper;
+
+    protected T findOne(String query, Object... params) {
+        try {
+            return jdbc.queryForObject(query, mapper, params);
+        } catch (EmptyResultDataAccessException ignored) {
+            return null;
+        }
```
Не очень хорошая практика игнорить исключения + возвращать null явно. Тут можно использовать метод query(), который не бросает исключение, если из БД не вернулось ничего

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/dal/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,109 @@
+            if (notHasGenreId) {
+                throw new NotFoundException("Genre с таким id не найдено");
+            }
+            film.getGenres().forEach(genre -> insert(ADD_FILM_GENRES, id, genre.getId()));
```
Запросы в цикле, для записи жанров можно в данном случае использовать batchUpdate() - запись пачками

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/dal/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,109 @@
+                film.getMpa().getId(),
+                film.getId()
+        );
+        film.getGenres().forEach(genre -> insert(ADD_FILM_GENRES, film.getId(), genre.getId()));
```
Тут тоже

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/dal/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,109 @@
+@Repository
+public class FilmDbStorage extends BaseStorage<Film> implements FilmStorage {
+
+    private static final String GET_FILMS = "SELECT * FROM films";
+    private static final String GET_FILM_BY_ID = "SELECT * FROM films WHERE film_id = ?";
```
С фильмами можно доставать сразу рейтинг, есть связь один к одному rating_id - mpa_id

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -17,32 +19,37 @@ public class FilmController {
     @PostMapping
-    public Film addFilm(@Valid @RequestBody Film film) {
-        return filmService.addFilm(film);
+    public Film addFilm(@Valid @RequestBody NewFilmRequest filmRequest) {
```
В `NewFilmRequest` нет аннотаций для валидации

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -17,32 +19,37 @@ public class FilmController {
     @PutMapping
-    public Film updateFilm(@Valid @RequestBody Film film) {
-        return filmService.updateFilm(film);
+    public Film updateFilm(@Valid @RequestBody UpdateFilmRequest request) {
```
В `UpdateFilmRequest` тоже

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
Раз в контроллеры приходят дтошки, то тут уже не нужна валидация, нужно перенести ее в дтошки

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -3,55 +3,94 @@
+    public Film addFilm(NewFilmRequest request) {
+        System.out.println("вошел в метод сервиса addFilm");
+        Film film = FilmMapper.mapToFilm(request);
+        System.out.println("прошел mapToFilm в метод сервиса addFilm");
```
Вместо печатей в консоль реализуй логи

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -3,55 +3,94 @@
-                .collect(Collectors.toList());
+        return filmStorage.getPopularFilms(count).stream()
+                .peek(film -> {
+                    film.setMpa(mpaStorage.getMpaById(film.getMpa().getId()));
+                    film.setGenres(genreStorage.getFilmGenres(film.getId()));
+                    film.setLikes(likesStorage.getFilmLikes(film.getId()));
```
Тоже запросы в цикле

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -3,55 +3,94 @@
+    public List<Film> getFilms() {
+        List<Film> films = filmStorage.getFilms();
+        for (Film film : films) {
+            film.setMpa(mpaStorage.getMpaById(film.getMpa().getId()));
+            film.setGenres(genreStorage.getFilmGenres(film.getId()));
+            film.setLikes(likesStorage.getFilmLikes(film.getId()));
```
Запросы в цикле, их нужно избегать, будет огромная нагрузка на БД при большом количестве данных (проблема N+1)

SergeyLeshchev replied:
получилось полностью убрать весь цикл, потому что теперь эти поля заполняются при маппинге :)

SergeyLeshchev committed:
Выполнить правки №1 к ТЗ №12

1. Убрать маппер LikeRowMapper
2. Реализовать маппинг лайков в LikeStorage
3. В BaseStorage изменить queryForObject на query
4. Заменить циклы с insert на один вызов batchUpdate
5. Изменить SQL запросы, чтобы фильмы приходили с mpa
6. Изменить маппинг фильмов, чтобы разгрузить FilmService
7. Изменить валидацию моделей и DTO
8. Убрать лишние строки(мусор) из кода

SergeyLeshchev committed:
Выполнить правки №1 к ТЗ №12

1. Убрать маппер LikeRowMapper
2. Реализовать маппинг лайков в LikeStorage
3. В BaseStorage изменить queryForObject на query
4. Заменить циклы с insert на один вызов batchUpdate
5. Изменить SQL запросы, чтобы фильмы приходили с mpa
6. Изменить маппинг фильмов, чтобы разгрузить FilmService
7. Изменить валидацию моделей и DTO
8. Убрать лишние строки(мусор) из кода

SergeyLeshchev committed:
Выполнить правки №1 к ТЗ №12

1. Удалить неиспользуемые импорты
2. Удалить mpaStorage из FilmService

SergeyLeshchev committed:
Выполнить правки №1 к ТЗ №12

1. Удалить неиспользуемые импорты
2. Удалить mpaStorage из FilmService

SergeyLeshchev committed:
Выполнить правки №1 к ТЗ №12

1. Добавить валидацию Mpa

SergeyLeshchev committed:
Убрать неиспользуемый импорт

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/dal/mappers/FilmRowMapper.java:
```diff
@@ -24,9 +26,11 @@ public Film mapRow(ResultSet resultSet, int rowNum) throws SQLException {
+                resultSet.getLong("mpa_id"),
+                resultSet.getString("mpa_name")));
+        film.setGenres(genreStorage.getFilmGenres(film.getId()));
+        film.setLikes(likesStorage.getFilmLikes(film.getId()));
```
Это такие же запросы в цикле, если мы запрашиваем все фильмы или подборку фильмов по лайкам, по каждой строке из БД мы идем отдельным запросом за жанрами и лайками. Плюс, внедрять стораджи в маппер тоже не лучшая идея, так как мапперы внедряются в стораджи и мы имеем шанс на зацикливание зависимостей. 

Самый очевидный вариант - в сервисе отдельно подгружать жанры и лайки (лайки можно получить и в виде количества вместе с самими фильмами) одним запросом

SergeyLeshchev committed:
Сделать правки №2 для ТЗ №12

1. Убрать вызовы получения жанров и лайков из маппинга Film
2. Добавить методы для получения жанров и лайков для списка фильмов
3. Добавлять жанры и лайки фильмам в сервисе
